icefoganalytics/travel-authorization
# Add Test Suite and Minimal Set Of Example Tests

Fixes https://github.com/icefoganalytics/travel-authorization/issues/31

Relates to:

- https://github.com/icefoganalytics/travel-authorization/issues/68

# Context

Add test suite and minimal set of example tests.
Required to solve https://github.com/icefoganalytics/travel-authorization/issues/68 in a sane manner.
Check out https://github.com/icefoganalytics/elcc-data-management for an example of how to do this.
And this PR https://github.com/icefoganalytics/elcc-data-management/pull/40/files

See https://github.com/thoughtbot/fishery/issues/120 for some typing help.

# Implementation

Split the `api/src/index.js` into `api/src/app.js` and `api/src/server.js` so you can test without booting the server.
Add "jest" for testing - if I wasn't using Typescript I would have gone with Mocha instead.
Add "ts-jest" for typescript support when testing.
Add "supertest" for request testing.
Add "fishery" for test factory support.
Silence a bunch of debug logging when running in "test" mode.
Fix bug in Estimate Generation thanks to having a test suite, there are still more bugs.
Add a bunch of test factories with semi-realistic data.

Note: the test environment gets it's own database at `travel_test`; Rails appends the environment to the database name, so I did that too.

# Screenshots

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/0a4c9198-b5e0-4124-a429-f9f998bfae9c)

# Testing Instructions

1. Boot the app via `dev up`
2. Log in to the app at http://localhost:8080
3. Now that you are confident that the app still runs, go back to your bash terminal and boot the test suite via
    ```bash
    dev test
    # or dev test_api
    ```
4. Check that the tests run, and that the only failing test is the "bulk generate" one.
